sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1571,6 +1659,162 @@ a normal SYCL command-group submission.
+resulting from this command-group submission is associated with.
+
+|===
+
+=== Dynamic Events [[dynamic-events]]
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+  class dynamic_event {
+    dynamic_event();
+    dynamic_event(const event& syclEvent);
+
+    void update(const event& syclEvent);
+  };
+}
+----
+
+Dynamic events represent SYCL events from outside of a given `command_graph`
+which nodes in that graph may depend on. These events are either obtained from
+normal SYCL operations or from another `command_graph` via `get_event()`. The
+`dynamic_event` object enables these dependent events to be updated between
+graph executions.
+
+Dynamic events can be used to add dependencies to a graph node in the same way
+that regular SYCL events can, by passing them as parameters to
+`handler::depends_on()` inside the CGF which represents the node.
+
+[source,c++]
+----
+// Obtain an event from a normal queue submission
+event OutsideEvent = queue.submit(...);
+
+// Create a dynamic event to wrap that event
+ext::oneapi::experimental::dynamic_event DynEvent {OutsideEvent};
+
+// Add a graph node which depends on that dynamic event
+Graph.add([&](handler& CGH){
+  CGH.depends_on(DynEvent);
+  CGH.parallel_for(...);
+});
+----
+Dynamic events created with an event from a `command_graph` cannot then be
+associated with other nodes in that same graph as this could be used change the
+topology of the graph. Attempting to call `handler::depends_on()` with such a
+`dynamic_event` in that situation will result in an error.
+
+Dynamic events can be created with no event but must be updated with a valid
+event before any executable graph which depends on that event is executed.
+Failing to do so will result in an error.
+
+The `dynamic_event` class provides the {crs}[common reference semantics].
+
+Table {counter: tableNumber}. Member functions of the `dynamic_event` class.
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+dynamic_event();
+----
+
+| Constructs a default `dynamic_event` which is not associated with any SYCL event.
+
+|
+[source,c++]
+----
+dynamic_event(const event& syclEvent);
+----
+
+| Constructs a `dynamic_event` which is associated with the SYCL event `syclEvent`.
+
+Parameters:
+
+* `syclEvent` - The SYCL event to associate this `dynamic_event` with.
+
+|
+[source,c++]
+----
+void update(const event& syclEvent);
+----
+
+| Updates the SYCL event associated with `dynamic_event`. This update will be
+reflected immediately in the associated modifiable graph nodes. An executable
+graph can then be updated to reflect these new event dependencies using
+<<executable-graph-update, Executable Graph Update>>.
+
+Parameters:
+
+* `syclEvent` - The new SYCL event to update this `dynamic_event` with.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if `syclEvent` is an event
+obtained from a queue submission which was recorded to the same graph any of the
```

Comment:
You could clarify this in the spec by using a consistent term to refer to these events, and put that term in italics.  Can we refer to these events as a _recorded edge_?  Then, you could rewrite this description like:

> Throws synchronously with error code `invalid` if `syclEvent` is a _recorded edge_.

Do we really need the rest of the description about being from the same graph?  Wouldn't it be a problem also if the event was a _recorded edge_ from a different graph?

Reply:
I think it would have to be a "recorded node" rather than "recorded edge", since I don't think you can equate one of these events to an edge conceptually. 

> Do we really need the rest of the description about being from the same graph? Wouldn't it be a problem also if the event was a recorded edge from a different graph?

Right now yes they should both be disallowed, I'll update the wording here. As discussed in [this comment](https://github.com/reble/llvm/pull/372#discussion_r1617576585) we may want to revisit that in the future to allow events from other graphs to avoid the need for manual updating of these inter-graph dependencies by the user.

Reply:
> so this wording should maybe be something like "if `syclEvent` represents a modifiable graph node".

This doesn't sound right to me, or maybe I don't understand the restriction.  I thought the restriction was only for commands that were added via a queue in record mode.  Doesn't the term "modifiable graph node" apply to nodes added through the explicit API?  Maybe you could write the restriction like this:

> Throws synchronously with error code `invalid` if `syclEvent` represents a _recorded command_.

However, I still think it would be better to define a term that applies to the event itself, rather than the command that it represents.  Therefore, maybe:

> Throws synchronously with error code `invalid` if `syclEvent` is a _recorded event_.


Separately:

> As discussed in https://github.com/reble/llvm/pull/372#discussion_r1617576585 we may want to revisit that in the future to allow events from other graphs to avoid the need for manual updating of these inter-graph dependencies by the user.

This sounds like it re-opens the discussion about "resetting" events.  The problem before was that we did not have a precise definition for when an event is reset.  I guess this is a discussion for another day, though.

Reply:
> This doesn't sound right to me, or maybe I don't understand the restriction.

The restriction is in place because these events returned from:

- A recorded queue submission or
- A hypothetical call to something like `graph.get_event_from_node(nodeObject);`

are not associated with backend events and are only used for dependency tracking within a graph. So events of this type from another graph are not allowed because there is nothing to create an actual backend dependency on and events from the same graph are not allowed because that is creating an edge within the graph (which cannot be updated).

Getting an event from a node is a gap in the interop functionality between our APIs which I do think is important to support, so I feel like "recorded" in this case is too API specific. Perhaps a "modifable-graph event"? To distinguish it from events obtained from an executable graph via `get_event()` which represent an actual execution.

> This sounds like it re-opens the discussion about "resetting" events.

I don't necessarily think it does since the idea discussed in that comment pertains to using the events detailed above to create these dependencies between graphs, and these events cannot be waited on by the user, so once the graphs are built these dependencies are fixed. Any resetting that occurs is purely an implementation detail, similar to what we already do with backend events in command-buffers right now, just extending that to have dependencies between command-buffers.

Reply:
> The restriction is in place because these events returned from:
> 
> * A hypothetical call to something like `graph.get_event_from_node(nodeObject);`
> 
> are not associated with backend events and are only used for dependency tracking within a graph

There is no function named `get_event_from_node`.  I assume you mean `command_graph::get_event`?  But the intended use of `get_event` is specifically to use it with `dynamic_event::update`.  This is even illustrated in the example below "Dynamic Event Update".